lib/eig_trlm_3d.cpp

```diff
@@ -0,0 +1,701 @@
+              logQuda(QUDA_DEBUG_VERBOSE, "**** Converged %d %d resid=%+.6e condition=%.6e ****\n", t, i,
+                      residua_3D[t][i + num_locked_3D[t]], tol * check_norm(alpha_3D[t][i + num_locked_3D[t]], t));
+              iter_converged_3D[t] = i+1;
+            } else {
+              // Unlikely to find new converged pairs
+              break;
+            }
+          }
+        }
+      }
+      
+      for (int t = 0; t < ortho_dim_size; t++) {
+        if (!converged_3D[t])
+          iter_keep_3D[t]
+            = std::min(iter_converged_3D[t] + (n_kr - num_converged_3D[t]) / 2, n_kr - num_locked_3D[t] - 12);
+      }
+
+      getProfile().TPSTOP(QUDA_PROFILE_COMPUTE);
+      computeKeptRitz3D(kSpace);
+      getProfile().TPSTART(QUDA_PROFILE_COMPUTE);
+
+      int min_nconv = n_kr+1;
+      int min_nlock = n_kr+1;
+      int min_nkeep = n_kr+1;
+
+      for (int t = 0; t < ortho_dim_size; t++) {
+        if (!converged_3D[t]) {
+          num_converged_3D[t] = num_locked_3D[t] + iter_converged_3D[t];
+          num_keep_3D[t] = num_locked_3D[t] + iter_keep_3D[t];
+          num_locked_3D[t] += iter_locked_3D[t];
+
+          if (num_converged_3D[t]<min_nconv) min_nconv=num_converged_3D[t];
+          if (num_locked_3D[t]<min_nlock) min_nlock=num_locked_3D[t];
+          if (num_keep_3D[t]<min_nkeep) min_nkeep=num_keep_3D[t];
+
+	  // Use printf to get data from t dim only
+          if (getVerbosity() >= QUDA_DEBUG_VERBOSE && comm_coord(0) == 0 && comm_coord(1) == 0 && comm_coord(2) == 0) {
+            printf("%04d converged eigenvalues for timeslice %d at restart iter %04d\n", num_converged_3D[t],
+                   t_offset + t, restart_iter + 1);
+            printf("iter Conv[%d] = %d\n", t_offset + t, iter_converged_3D[t]);
+            printf("iter Keep[%d] = %d\n", t_offset + t, iter_keep_3D[t]);
+            printf("iter Lock[%d] = %d\n", t_offset + t, iter_locked_3D[t]);
+            printf("num_converged[%d] = %d\n", t_offset + t, num_converged_3D[t]);
+            printf("num_keep[%d] = %d\n", t_offset + t, num_keep_3D[t]);
+            printf("num_locked[%d] = %d\n", t_offset + t, num_locked_3D[t]);
+            for (int i = 0; i < n_kr; i++) {
+              printf("Ritz[%d][%d] = %.16e residual[%d] = %.16e\n", t_offset + t, i, alpha_3D[t][i], i,
+                     residua_3D[t][i]);
+            }
+          }
+        }
+      }
+
+      // Check for convergence
+      bool all_converged = true;
+      for (int t = 0; t < ortho_dim_size; t++) {
+        if (num_converged_3D[t] >= n_conv) {
+          converged_3D[t] = true; 
+        } else {
+          all_converged = false;
+        }
+      }
+
+      if (getVerbosity() >= QUDA_VERBOSE && comm_coord(0) == 0 && comm_coord(1) == 0 && comm_coord(2) == 0) {
+        printf("iter = %d rank = %d converged = ", restart_iter + 1, comm_rank_global());
```

Comment:
I see what's going on here, do we want to make a custom `printfQuda` (`printfQuda3d`?) that handles this type of conditional on the print instead of using `printf`? Or should we avoid over-specializing?

In any case---it might be good to modify this to build the entire string before printing so the outputs from different ranks don't get interleaved... or at least minimize the risk.

Reply:
I'd steer away from `printfQuda3d` but your point about building the entire string first is a good one.

Reply:
030775807